File: CSharp/ByTier/B/1907_B_YetnotherrokenKeoard.cs
// Problem: CF 1907 B - YetnotherrokenKeoard
// https://codeforces.com/contest/1907/problem/B

/*
 * Problem: YetAnotherBrokenKeyboard
 * Purpose: Simulate a broken keyboard where 'b' deletes the last lowercase letter and 'B' deletes the last uppercase letter.
 *          The result is the final string after all key presses are processed.
 * 
 * Algorithms/Techniques:
 *   - Use two lists to store indices of lowercase and uppercase letters typed so far.
 *   - When 'b' or 'B' is pressed, decrement the corresponding index (simulate deletion).
 *   - After processing all presses, reconstruct the string by merging the remaining lowercase and uppercase letters
 *     in the order they were typed (maintaining their relative positions).
 * 
 * Time Complexity: O(n) where n is the total length of all input strings.
 * Space Complexity: O(n) for storing indices and result string.
 */

using System;
using System.Collections.Generic;
using System.Text;

public class YetnotherrokenKeoard {
  public static void Main() {
    short t = short.Parse(Console.ReadLine());
    for (short i = 0; i < t; i++) {
      string s = Console.ReadLine();
      List<int> lower = new List<int>();  // Store indices of lowercase letters
      List<int> upper = new List<int>();  // Store indices of uppercase letters
      int idxLower = 0;                   // Current position in lower list
      int idxUpper = 0;                   // Current position in upper list
      int sLen = s.Length;
      for (int j = 0; j < sLen; j++) {
        // Process 'b' key - delete last lowercase letter
        if (s[j] == 'b')
          idxLower = Math.Max(0, idxLower - 1);
        // Process 'B' key - delete last uppercase letter
        else if (s[j] == 'B')
          idxUpper = Math.Max(0, idxUpper - 1);
        // Process lowercase letter - add to lower list
        else if (char.IsLower(s[j])) {
          if (idxLower < lower.Count)
            lower[idxLower] = j;  // Reuse existing index if possible
          else
            lower.Add(j);         // Add new index
          idxLower++;
        } else {
          // Process uppercase letter - add to upper list
          if (idxUpper < upper.Count)
            upper[idxUpper] = j;  // Reuse existing index if possible
          else
            upper.Add(j);         // Add new index
          idxUpper++;
        }
      }
      // Build final output string by merging lower and upper letters in order
      StringBuilder ans = new StringBuilder();
      int il = 0;  // Pointer for lower list
      int iu = 0;  // Pointer for upper list
      while (il < idxLower && iu < idxUpper) {
        // Pick the next character based on which was typed earlier
        if (lower[il] < upper[iu]) {
          ans.Append(s[lower[il]]);
          il++;
        } else {
          ans.Append(s[upper[iu]]);
          iu++;
        }
      }
      // Append remaining lowercase letters
      while (il < idxLower) {
        ans.Append(s[lower[il]]);
        il++;
      }
      // Append remaining uppercase letters
      while (iu < idxUpper) {
        ans.Append(s[upper[iu]]);
        iu++;
      }
      Console.WriteLine(ans);
    }
  }
}


// https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/B/1907_B_YetnotherrokenKeoard.cpp
// Problem: CF 1907 B - YetnotherrokenKeoard
// https://codeforces.com/contest/1907/problem/B

/*
 * Problem: Yet Another Broken Keyboard
 * 
 * Purpose:
 *   Simulates a broken keyboard where pressing 'b' deletes the last lowercase letter
 *   and pressing 'B' deletes the last uppercase letter from the typed string.
 *   The letters 'b' and 'B' themselves are not added to the string.
 *
 * Algorithm:
 *   - Use two stacks (vectors) to track indices of lowercase and uppercase letters.
 *   - Process each character in the input string:
 *     * If 'b': delete last lowercase letter (pop from smalls vector)
 *     * If 'B': delete last uppercase letter (pop from caps vector)
 *     * If lowercase letter: push index onto smalls
 *     * If uppercase letter: push index onto caps
 *   - Mark characters to be shown using a boolean vector.
 *   - Output remaining characters in order.
 *
 * Time Complexity: O(n), where n is the length of the input string (each character processed once).
 * Space Complexity: O(n), for storing indices and the boolean show vector.
 */

#include <iostream>
#include <vector>
#include <string>

int main() {
  std::ios_base::sync_with_stdio(false);
  long t;
  std::cin >> t;
  while (t--) {
    std::string s;
    std::cin >> s;
    std::vector<long> smalls, caps;           // Stack to store indices of lowercase and uppercase letters
    std::vector<bool> show(s.size(), true);   // Boolean vector to mark which characters to show
    for (long p = 0; p < s.size(); p++) {
      if (s[p] == 'b') {
        show[p] = false;                      // Mark 'b' as not shown
        if (smalls.size() == 0) {             // If no lowercase letters, ignore
          continue;
        }
        show[smalls.back()] = false;          // Mark last lowercase letter to be hidden
        smalls.pop_back();                    // Remove from stack
      } else if (s[p] == 'B') {
        show[p] = false;                      // Mark 'B' as not shown
        if (caps.size() == 0) {               // If no uppercase letters, ignore
          continue;
        }
        show[caps.back()] = false;            // Mark last uppercase letter to be hidden
        caps.pop_back();                      // Remove from stack
      } else if (s[p] >= 'a' && s[p] <= 'z') {
        smalls.push_back(p);                  // Push lowercase letter index
      } else if (s[p] >= 'A' && s[p] <= 'Z') {
        caps.push_back(p);                    // Push uppercase letter index
      }
    }
    for (long p = 0; p < s.size(); p++) {
      if (show[p]) {                        // If character is marked to be shown
        std::cout << s[p];
      }
    }
    std::cout << std::endl;
  }
}


// https://github.com/VaHiX/codeForces/